src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        # parameters takes. It can be assumed that it's fast --> wait 1 s.
+        time.sleep(1)
+
+    def ValidateChannels(self):
+        """
+        Validate the channels. This function is blocking.
+        This is a quick adjustment routine which needs to be performed after loading the configuration.
+        """
+        # There is also a channel validation progress event, but it's triggered too early. We need
+        # to wait until the state becomes stable again before continuing.
+        self.EnableEventNotification(SA_SIDLL.SA_PS_STABLE_STATE_CHANGED_EVENT)
+        self.SetProperty_i32(SA_SIDLL.SA_PS_AF_CHANNEL_VALIDATION_STATE_PROP, SA_SIDLL.SA_SI_ENABLED)
+        self.WaitForEvent(SA_SIDLL.SA_SI_ENABLED)
+        self.DisableEventNotification(SA_SIDLL.SA_PS_STABLE_STATE_CHANGED_EVENT)
+
+    def EnableChannel(self, channel):
+        """
+        Enable channel. This function is blocking.
+        channel (int): channel number
+        """
+        self.SetProperty_i32(SA_SIDLL.SA_PS_CH_ENABLED_PROP, SA_SIDLL.SA_SI_ENABLED, idx0=channel)
+        # Channel enabled event not triggered if it was already enabled
+        # --> poll SA_PS_CH_ENABLED_PROP attribute instead
+        while self.GetProperty_i32(SA_SIDLL.SA_PS_CH_ENABLED_PROP, idx0=channel) != SA_SIDLL.SA_SI_ENABLED:
+            time.sleep(0.1)
+
+    def EnableEventNotification(self, event_type):
+        """
+        Enables notifications for an event type.
+        event_type: (int) SA_SIDLL event type
+        """
+        encoded_key = SA_SIDLL.SA_SI_EVENT_NOTIFICATION_ENABLED_PROP << 16 | event_type
+        self.core.SA_SI_SetProperty_i32(self._id, c_uint32(encoded_key), c_int32(SA_SIDLL.SA_SI_ENABLED))
+
+    def DisableEventNotification(self, event_type):
+        """
+        Disables notifications for an event type.
+        event_type: (int) SA_SIDLL event type
+        """
+        encoded_key = SA_SIDLL.SA_SI_EVENT_NOTIFICATION_ENABLED_PROP << 16 | event_type
+        self.core.SA_SI_SetProperty_i32(self._id, c_uint32(encoded_key), c_int32(SA_SIDLL.SA_SI_DISABLED))
+
+    def WaitForEvent(self, end_state, timeout=float("inf")):
+        """
+        Blocks until event is triggered or timeout.
+        end_state (int): state of event variable to wait for
+        timeout (float): maximum time to wait in s. If inf, it will wait forever.
+        """
+        if timeout == float("inf"):
+            t = SA_SIDLL.SA_SI_TIMEOUT_INFINITE
+        else:
+            t = c_uint(int(timeout * 1000))
+        ev = SA_SI_Event()
+
+        state = None
+        while state != end_state:
+            # Don't check event type, it's not always accurate.
+            # Instead, only one event is active and all other events are disabled, so it is clear
+            # which event we are waiting for.
+            self.core.SA_SI_WaitForEvent(self._id, byref(ev), t)
+            state = ev.devEventParameter & 0xffff
+            time.sleep(0.1)
```

Comment:
I don't think you need to have a sleep here, as WaitForEvent already blocks.
However, you need to check if the timeout is passed.

Reply:
I think I put it there for a purpose, just waiting for the end_state was sometimes not enough. But I should check that again.